### PR TITLE
Fix retry code to hold refs to send_initial_metadata slices

### LINKED ISSUE #26994: Retry routine causes use-after-free segfault in PHP extension
### What version of gRPC and what language are you using?
1.39.0, PHP

### What operating system (Linux, Windows,...) and version?
Linux, MacOS

### What runtime / compiler are you using (e.g. python version or version of gcc)
Reproducible on both clang and gcc, php 7.4 and 8.0

### What did you do?

gRPC Has a cool feature - it can re-try requests on a particular set of response-codes.
It looks like this retry is called asynchronously via timer and grpc_completion_queue_pluck() doesn't wait until retries are done. This results in metadata cleanup which is supposed to be used on retry (and it feels like, PHP flow executes before gRPC call is actually done).

Steps to reproduce:

1. Setup channel to retry on status UNAVAILABLE (see code below, "retryableStatusCodes" setting in the code)
2. Setup a gRPC service to return UNAVAILABLE on each request
3. Perform the code (it is essential to perform the call twice)

```php
$credentials = \Grpc\ChannelCredentials::createInsecure();

$data = [
    'sapi-user-bin' =>
        [
            0 => \base64_decode('CMPz7eADEAEY/gEgASgWMDc4qwZA7rJSUAFYgAJiAGgD'),
        ],
    'sapi-appinfo-bin' =>
        [
            0 => \base64_decode('CgoIAhABGAVKAhABEp0BCiQzMTY4Yjc5OC1iNzk4LTk4NmEtNmEyNy0yNzRlOTFhM2MyMWMiBWVuLUdCKgsKBU1hY09TEAoYDzJhClJNb3ppbGxhLzUuMCAoTWFjaW50b3NoOyBJbnRlbCBNYWMgT1MgWCAxMC4xNTsgcnY6OTAuMCkgR2Vja28vMjAxMDAxMDEgRmlyZWZveC85MC4wEgsKB0ZJUkVGT1gQWhoxczI1NDoxMTI6M21KYzZUYWJXV3hwbnF6c1paa0liMlJDN1l5Y1hjYzhnMThiYlZoaCABKP8BMiRjZGM1MDc0Ny1jOTFmLTQ0ZjAtOGY3YS1kZjkyZjgxYzBiYzZAaUj/AQ=='),
        ],
    'sapi-conn-bin' =>
        [
            0 => \base64_decode('CgR/AAABEgoKCDM2MzU2NDU2'),
        ],
];

$clientOptions = array (
    'credentials' => $credentials,
    'grpc.lb_policy_name' => 'round_robin',
    'grpc.max_connection_age_ms' => 300000,
    'grpc.max_connection_age_grace_ms' => 30000,
    'grpc.keepalive_time_ms' => 10000,
    'grpc.keepalive_timeout_ms"' => 5000,
    'grpc.service_config' => '{"loadBalancingPolicy":"round_robin","methodConfig":[{"name":[{"service":"com.bumble.contract.service.qapier.QaPierService","method":"MethodList"},{"service":"com.bumble.contract.service.interests.InterestsService","method":"GetUser"},{"service":"com.bumble.contract.service.interests.InterestsService","method":"SaveUserInterests"}],"waitForReady":true,"timeout":"5s","retryPolicy":{"maxAttempts":2,"initialBackoff":"0.9s","maxBackoff":"2.5s","backoffMultiplier":2,"retryableStatusCodes":["UNAVAILABLE","RESOURCE_EXHAUSTED","DEADLINE_EXCEEDED"]}}],"retryThrottling":{"maxTokens":10,"tokenRatio":0.1}}',
    'grpc.enable_retries' => 1,
);

$Client = new \BumbleApi\Service\Qapier\QaPierServiceClient(

    'localhost:50051',
    ['credentials' => $credentials] + $clientOptions
);

[$response, $status] = $Client->MethodList(new \BumbleApi\Service\Qapier\MethodListRequest(), $data, ['request_timeout' => 900000.0])->wait();

var_dump($status);

[$response, $status] = $Client->MethodList(new \BumbleApi\Service\Qapier\MethodListRequest(), $data, ['request_timeout' => 900000.0])->wait();
var_dump($status);

```

ASAN report:
```
$ DYLD_INSERT_LIBRARIES=/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/lib/darwin/libclang_rt.asan_osx_dynamic.dylib php -dgrpc.enable_fork_support=1 -dgrpc.poll_strategy=poll -dextension=./grpc.so -dgrpc.grpc_verbosity=debug -dgrpc.grpc_trace=all -dgrpc.log_filename=tmp.log test.php
object(stdClass)#541 (3) {
  ["metadata"]=>
  array(0) {
  }
  ["code"]=>
  int(14)
  ["details"]=>
  string(91) "upstream connect error or disconnect/reset before headers. reset reason: connection failure"
}
 =================================================================
==20786==ERROR: AddressSanitizer: heap-use-after-free on address 0x61200001ce40 at pc 0x00010a0f4f0b bp 0x7000091d4d40 sp 0x7000091d4d38
READ of size 8 at 0x61200001ce40 thread T3
    #0 0x10a0f4f0a in maybe_link_callout(grpc_metadata_batch*, grpc_linked_mdelem*) metadata_batch.cc:130
    #1 0x10a0f5e65 in grpc_metadata_batch_link_tail(grpc_metadata_batch*, grpc_linked_mdelem*) metadata_batch.cc:231
    #2 0x10a0f5dc0 in grpc_metadata_batch_add_tail(grpc_metadata_batch*, grpc_linked_mdelem*, grpc_mdelem) metadata_batch.cc:209
    #3 0x10a0f9ba7 in grpc_metadata_batch_copy(grpc_metadata_batch*, grpc_metadata_batch*, grpc_linked_mdelem*) metadata_batch.cc:405
    #4 0x1098836d2 in grpc_core::(anonymous namespace)::RetryFilter::CallData::CallAttempt::BatchData::AddRetriableSendInitialMetadataOp() retry_filter.cc:1804
    #5 0x109881371 in grpc_core::(anonymous namespace)::RetryFilter::CallData::CallAttempt::MaybeCreateBatchForReplay() retry_filter.cc:772
    #6 0x10988104f in grpc_core::(anonymous namespace)::RetryFilter::CallData::CallAttempt::AddRetriableBatches(grpc_core::CallCombinerClosureList*) retry_filter.cc:962
    #7 0x1098717ba in grpc_core::(anonymous namespace)::RetryFilter::CallData::CallAttempt::StartRetriableBatches() retry_filter.cc:979
    #8 0x1098714d0 in grpc_core::(anonymous namespace)::RetryFilter::CallData::CreateCallAttempt() retry_filter.cc:2140
    #9 0x10988a601 in grpc_core::(anonymous namespace)::RetryFilter::CallData::OnRetryTimerLocked(void*, grpc_error*) retry_filter.cc:2426
    #10 0x109de662f in exec_ctx_run(grpc_closure*, grpc_error*) exec_ctx.cc:43
    #11 0x109de618a in grpc_core::ExecCtx::Flush() exec_ctx.cc:165
    #12 0x109e5100c in run_some_timers() timer_manager.cc:134
    #13 0x109e50bde in timer_main_loop() timer_manager.cc:237
    #14 0x109e509c2 in timer_thread(void*) timer_manager.cc:284
    #15 0x109da946f in grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::'lambda'(void*)::operator()(void*) const thd_posix.cc:140
    #16 0x109da8df4 in grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::'lambda'(void*)::__invoke(void*) thd_posix.cc:110
    #17 0x7fff205af8fb in _pthread_start+0xdf (libsystem_pthread.dylib:x86_64+0x68fb)
    #18 0x7fff205ab442 in thread_start+0xe (libsystem_pthread.dylib:x86_64+0x2442)

0x61200001ce40 is located 0 bytes inside of 312-byte region [0x61200001ce40,0x61200001cf78)
freed by thread T0 here:
    #0 0x1022c8609 in wrap_free+0xa9 (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x48609)
    #1 0x109d65a99 in gpr_free alloc.cc:52
    #2 0x10a08d81f in grpc_metadata_array_destroy metadata_array.cc:35
    #3 0x10a186ed9 in grpc_php_metadata_array_destroy_including_entries call.c:177
    #4 0x10a18af68 in zim_Call_startBatch call.c:529
    #5 0x1017c971c in ZEND_DO_FCALL_SPEC_RETVAL_UNUSED_HANDLER+0x150 (php:x86_64+0x10034371c)
    #6 0x1017ae591 in execute_ex+0x22 (php:x86_64+0x100328591)
    #7 0x1017ae75a in zend_execute+0x16e (php:x86_64+0x10032875a)
    #8 0x101794a41 in zend_execute_scripts+0x11b (php:x86_64+0x10030ea41)
    #9 0x10173b64c in php_execute_script+0x171 (php:x86_64+0x1002b564c)
    #10 0x1018213a5 in do_cli+0x17fe (php:x86_64+0x10039b3a5)
    #11 0x10181fa4d in main+0x581 (php:x86_64+0x100399a4d)
    #12 0x7fff205caf5c in start+0x0 (libdyld.dylib:x86_64+0x15f5c)

previously allocated by thread T0 here:
    #0 0x1022c84c0 in wrap_malloc+0xa0 (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x484c0)
    #1 0x109d659d8 in gpr_malloc alloc.cc:32
    #2 0x10a186090 in create_metadata_array call.c:141
    #3 0x10a189085 in zim_Call_startBatch call.c:340
    #4 0x1017c971c in ZEND_DO_FCALL_SPEC_RETVAL_UNUSED_HANDLER+0x150 (php:x86_64+0x10034371c)
    #5 0x1017ae591 in execute_ex+0x22 (php:x86_64+0x100328591)
    #6 0x1017ae75a in zend_execute+0x16e (php:x86_64+0x10032875a)
    #7 0x101794a41 in zend_execute_scripts+0x11b (php:x86_64+0x10030ea41)
    #8 0x10173b64c in php_execute_script+0x171 (php:x86_64+0x1002b564c)
    #9 0x1018213a5 in do_cli+0x17fe (php:x86_64+0x10039b3a5)
    #10 0x10181fa4d in main+0x581 (php:x86_64+0x100399a4d)
    #11 0x7fff205caf5c in start+0x0 (libdyld.dylib:x86_64+0x15f5c)

Thread T3 created by T0 here:
    #0 0x1022c25ea in wrap_pthread_create+0x5a (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x425ea)
    #1 0x109da87d6 in grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&) thd_posix.cc:108
    #2 0x109da804c in grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&) thd_posix.cc:76
    #3 0x109da7c67 in grpc_core::Thread::Thread(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&) thd_posix.cc:186
    #4 0x109da809c in grpc_core::Thread::Thread(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&) thd_posix.cc:184
    #5 0x109e50809 in start_timer_thread_and_unlock() timer_manager.cc:92
    #6 0x109e50028 in start_threads() timer_manager.cc:293
    #7 0x109e4fffa in grpc_timer_manager_init() timer_manager.cc:311
    #8 0x109df51c8 in grpc_iomgr_start() iomgr.cc:69
    #9 0x10a089ae6 in grpc_init init.cc:164
    #10 0x10a19b15e in zm_activate_grpc php_grpc.c:599
    #11 0x10179b137 in zend_activate_modules+0x29 (php:x86_64+0x100315137)
    #12 0x101739bbb in php_request_startup+0x208 (php:x86_64+0x1002b3bbb)
    #13 0x101820e32 in do_cli+0x128b (php:x86_64+0x10039ae32)
    #14 0x10181fa4d in main+0x581 (php:x86_64+0x100399a4d)
    #15 0x7fff205caf5c in start+0x0 (libdyld.dylib:x86_64+0x15f5c)

SUMMARY: AddressSanitizer: heap-use-after-free metadata_batch.cc:130 in maybe_link_callout(grpc_metadata_batch*, grpc_linked_mdelem*)
Shadow bytes around the buggy address:
  0x1c2400003970: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c2400003980: fd fd fd fd fd fd fd fd fd fd fa fa fa fa fa fa
  0x1c2400003990: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x1c24000039a0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c24000039b0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fa fa
=>0x1c24000039c0: fa fa fa fa fa fa fa fa[fd]fd fd fd fd fd fd fd
  0x1c24000039d0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c24000039e0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fa
  0x1c24000039f0: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x1c2400003a00: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c2400003a10: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==20786==ABORTING
Abort trap: 6

```

### PR DESCRIPTION
**Pull request by markdroth**
Fixes #26994.

We've run into a problem whereby the retry code was not properly holding refs to the send_initial_metadata payload, so if the code above C-core destroyed the metadata after the send_initial_metadata op was completed, the retry code would crash when it tried to replay that op on a subsequent call attempt.

There are two parts to this problem.  First, metadata is passed in via the C-core API as an array of `grpc_metadata` structs.  When this is converted to a `grpc_metadata_batch` via `grpc_mdelem_from_grpc_metadata()`, the resulting mdelem [is of storage type `GRPC_MDELEM_STORAGE_EXTERNAL`](https://github.com/grpc/grpc/blob/0407f4be49d5754bbc2e5bc011c55b58e17d4f42/src/core/lib/transport/metadata.cc#L381), which is not ref-counted.  Thus, when the retry code called `grpc_metadata_batch_copy()` to make a copy of the batch, the copy was [not actually holding a new ref](https://github.com/grpc/grpc/blob/0407f4be49d5754bbc2e5bc011c55b58e17d4f42/src/core/lib/transport/metadata.h#L341).   To fix this, I have changed `grpc_metadata_batch_copy()` to explicitly create a new mdelem instead of reffing the original one, if the original one is not a ref-counted storage type.

Second, the optimization added in #19427 made it so that slices created via `grpc_slice_from_copied_string()` were not actually ref-counted as they should have been.  Thus, even if the mdelem was holding a ref to those slices, that ref would have been a no-op.  To fix this, I have changed `grpc_slice_from_copied_string()` to allocate a new ref-counted slice, the way it did prior to #19427.  This change may have a performance impact, but it's expected that future improvements can compensate for this.

@lidizheng and @stanley-cheung, please try out this fix and verify that it solves the PHP retry crashing problem.

### CODE REVIEW

File: src/core/lib/slice/slice.cc [hunk 1]
```diff
@@ -223,11 +223,15 @@ grpc_core::UnmanagedMemorySlice::UnmanagedMemorySlice(const char* source)
                                                             strlen(source)) {}
 
 grpc_slice grpc_slice_from_copied_buffer(const char* source, size_t length) {
-  return grpc_core::UnmanagedMemorySlice(source, length);
```
[the remark this answers is not in the document]
Comment (markdroth) [reply src/core/lib/slice/slice.cc:226]: Right, it doesn't use inlining.  As far as I can tell, the structure here provides no way to ref-count without allocating.

File: src/core/lib/transport/metadata_batch.cc [hunk 2]
```diff
@@ -401,9 +401,21 @@ void grpc_metadata_batch_copy(grpc_metadata_batch* src,
   size_t i = 0;
   for (grpc_linked_mdelem* elem = src->list.head; elem != nullptr;
        elem = elem->next) {
+    // If the mdelem is ref-counted, take a ref.
+    // Otherwise, create a new copy, holding its own refs to the
+    // underlying slices.
+    grpc_mdelem md;
+    if (GRPC_MDELEM_STORAGE(elem->md) == GRPC_MDELEM_STORAGE_INTERNED ||
```
[the remark this answers is not in the document]
Comment (markdroth) [reply src/core/lib/transport/metadata_batch.cc]: Done.

File: src/core/lib/slice/slice.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (ctiller) [reply src/core/lib/slice/slice.cc:226]: Right, but inlined slices don't have refcounting, but are completely owned.

[the remark this answers is not in the document]
Comment (markdroth) [reply src/core/lib/slice/slice.cc:226]: Doh!  Sorry, yes, I should have realized that.

Fixed.

**Review by ctiller — APPROVED** [review]

**Review by arjunroy — APPROVED** [review]
Looks fine on my end.

**Pull request merged**